fableous-be/handlers/classroom.go

```diff
@@ -48,3 +63,17 @@ func (m *module) ClassroomUpdate(classroomInfo datatransfers.ClassroomInfo) (err
 	}
 	return
 }
+
+func (m *module) ClassroomDeleteByID(classroomID string) (err error) {
```

Comment:
thought of a qs, is deleting a classroom when there is an ongoing session allowed? if yes, is behavior already as expected?

on another note, is deleting a classroom a superset of whats done when deleting a session? if so, just for consideration, should ClassroomDeleteByID() call SessionDeleteByIDByClassroomID()? or should both be called on controller.DELETEClassroom()? or they are fine as is

Reply:
yes, as we have discussed, it is allowed to delete classroom with ongoing session. 

i think, pattern-wise, we shouldnt call both in the controller. despite ClassroomDelete, by behaviour is a superset of SessionDelete, one shouldnt call SessionDeleteByIDByClassroomID() from ClassroomDeleteByID(). due to the current implementation -- the activeSession map uses classroomToken for its keys -- this session map has to be iterated to perform cleanup on any of the matching session, thus O(n) time. if we were to call SessionDeleteByIDByClassroomID(), the action would lead to O(n^2) runtime, and this is not an acceptable running time complexity given that the map would be read-locked throughout the delete operation.

any ideas for further improvements?

Reply:
ah okay nice

right okay, perhaps extracting logic to SessionCleanUp() like u already did is enough, not much duplication is left anyways

upon rereading the code, thought of 2 more things to clarify:

- in ClassroomDeleteByID(), do we need to call sessionOrmer.delete too if theres an ongoing session?
- I notice in ClassroomDeleteByID(), you call os.RemoveAll() and `go SessionCleanUp()`, I assume if the classroom-level directory is deleted first before the session-level directory, there will be no errors?

Reply:
- FK for ClassroomID in Session table is set to CASCADE ondelete, so i dont think its necessary
- the deletion of the session folders in SessionCleanUp() is executed in a separate thread. there is no absolute certainty on which one will be executed first: a) deletion of parent classroom folder, or; b) deleteion of session subfolders. however, at the end of all executions, all folders will eventually be deleted anyways, and if there errors (if any) returned by os.RemoveAll() in SessionCleanUp(), it will be ignored.